File: scripts/kb_pipeline/dedup.py
"""Deduplication — URL matching + embedding similarity + content-hash gate."""

import hashlib
import json
import logging
import re
import urllib.error
import urllib.request
from typing import Optional, List, Tuple

from .store import check_url_exists
from .config import DEDUP_SIMILARITY_THRESHOLD, get_supabase_url, get_supabase_secret_key

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Content-hash dedup gate (S183 WP2)
# ---------------------------------------------------------------------------

# Mirror of `lib/dedup.ts::DEDUP_MIN_CONTENT_LENGTH`. Content shorter than
# this (post-normalisation) collides too easily; callers should skip the
# hash check below this threshold. Reference:
# docs/specs/cross-system-dedup-spec.md §6 Risks.
DEDUP_MIN_CONTENT_LENGTH = 50

_NON_WORD = re.compile(r"[^\w\s]")
_WHITESPACE = re.compile(r"\s+")


def normalise_content_for_hash(text: str) -> str:
    """Normalise text for md5 hashing.

    Matches the PostgreSQL generated column `content_items.content_text_hash`
    byte-for-byte so a client-side hash is guaranteed to match the stored
    server-side hash. Matches `lib/dedup.ts::normaliseTextForHash`.

    Pipeline: lowercase -> trim -> strip non-word/non-space -> collapse
    whitespace -> trim.
    """
    if not text:
        return ""
    out = text.lower().strip()
    out = _NON_WORD.sub("", out)
    out = _WHITESPACE.sub(" ", out)
    return out.strip()


def check_content_hash_duplicate(content_text: str) -> Tuple[bool, Optional[str]]:
    """Shared dedup gate — returns (is_duplicate, existing_item_id) on match.

    Computes md5(normalised(text)) client-side and calls the
    `find_exact_duplicates` RPC. Returns (False, None) for content below
    the length threshold, a network failure, or no match. Never raises.

    Callers in Python ingest scripts should proceed with the insert on
    match and set `dedup_status='suspected_duplicate'` per S182b Q1
    (soft block). Reference: docs/specs/cross-system-dedup-spec.md §3.2.
    """
    if not content_text:
        return False, None

    normalised = normalise_content_for_hash(content_text)
    if len(normalised) < DEDUP_MIN_CONTENT_LENGTH:
        return False, None

    content_hash = hashlib.md5(normalised.encode("utf-8")).hexdigest()

    rpc_url = f"{get_supabase_url()}/rest/v1/rpc/find_exact_duplicates"
    body = json.dumps({"p_content_hash": content_hash}).encode("utf-8")

    key = get_supabase_secret_key()
    req = urllib.request.Request(rpc_url, data=body, method="POST")
    req.add_header("apikey", key)
    req.add_header("Authorization", f"Bearer {key}")
    req.add_header("Content-Type", "application/json")

    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            if not data:
                return False, None
            first = data[0] if isinstance(data, list) else data
            existing_id = first.get("id") if isinstance(first, dict) else None
            return (existing_id is not None), existing_id
    except (urllib.error.URLError, urllib.error.HTTPError, OSError) as e:
        logger.warning("Content-hash dedup check failed (network): %s", e)
        return False, None
    except (json.JSONDecodeError, KeyError, ValueError) as e:
        logger.warning("Content-hash dedup check failed (data): %s", e)
        return False, None


def check_duplicate_url(source_url: str) -> Optional[str]:
    """Check if URL already exists. Returns existing item ID or None."""
    if not source_url:
        return None
    return check_url_exists(source_url)


def check_duplicate_embedding(
    embedding: List[float],
    threshold: float = DEDUP_SIMILARITY_THRESHOLD,
    limit: int = 3,
) -> List[dict]:
    """Check for similar content via embedding similarity.

    Returns list of matches above threshold: [{id, title, similarity}]
    """
    key = get_supabase_secret_key()

    # Use the find_similar_content function via RPC
    rpc_url = f"{get_supabase_url()}/rest/v1/rpc/find_similar_content"
    body = json.dumps({
        "query_embedding": embedding,
        "similarity_threshold": threshold,
        "limit_count": limit,
    }).encode("utf-8")

    req = urllib.request.Request(rpc_url, data=body, method="POST")
    req.add_header("apikey", key)
    req.add_header("Authorization", f"Bearer {key}")
    req.add_header("Content-Type", "application/json")

    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            return [
                {
                    "id": item.get("id"),
                    "title": item.get("title", ""),
                    "similarity": item.get("similarity", 0),
                }
                for item in data
                if item.get("similarity", 0) >= threshold
            ]
    except (urllib.error.URLError, urllib.error.HTTPError, OSError) as e:
        logger.warning("Dedup embedding check failed (network): %s", e)
        return []
    except (json.JSONDecodeError, KeyError, ValueError) as e:
        logger.warning("Dedup embedding check failed (data): %s", e)
        return []


def is_duplicate(
    source_url: str = "",
    embedding: List[float] = None,
    threshold: float = DEDUP_SIMILARITY_THRESHOLD,
) -> tuple[bool, Optional[str], str]:
    """Check if content is a duplicate.

    Returns (is_dup, existing_id, method) where method is 'url' or 'embedding'.
    """
    # URL match first (fast)
    existing_id = check_duplicate_url(source_url)
    if existing_id:
        return True, existing_id, "url"

    # Embedding similarity (slower)
    if embedding:
        matches = check_duplicate_embedding(embedding, threshold)
        if matches:
            return True, matches[0]["id"], "embedding"

    return False, None, ""

File: lib/dedup.ts
/**
 * Content deduplication utilities for web upload and manual creation pathways.
 *
 * Two detection strategies:
 * 1. Exact match: MD5 hash of normalised content text (computed in SQL)
 * 2. Near-duplicate: Embedding cosine similarity via the find_similar_content RPC
 *
 * Both return informational warnings — they do not block content creation.
 */

import type { SupabaseClient } from '@supabase/supabase-js';
import type { Database } from '@/supabase/types/database.types';
import { logger } from '@/lib/logger';

/** Default cosine similarity threshold for near-duplicate detection */
export const DEFAULT_NEAR_DUPLICATE_THRESHOLD = 0.92;

/** Maximum number of near-duplicate matches to return */
const MAX_NEAR_DUPLICATE_RESULTS = 5;

export interface DuplicateMatch {
  id: string;
  title: string;
  similarity: number;
  match_type: 'exact' | 'near_duplicate';
}

export interface DedupResult {
  has_duplicates: boolean;
  matches: DuplicateMatch[];
}

/**
 * Minimum content length (post-normalisation, pre-hash) before we apply
 * content-hash dedup. Short strings (e.g. a 20-word Q&A) collide too
 * easily after punctuation/whitespace stripping. Below the threshold
 * callers should fall through to title-norm or skip dedup entirely.
 * Reference: cross-system-dedup-spec.md §6 Risks.
 */
export const DEDUP_MIN_CONTENT_LENGTH = 50;

/** Outcome of the shared dedup gate used by every TS entry point. */
/** @public */
export interface ExactDuplicateCheck {
  isDuplicate: boolean;
  existingId?: string;
  existingTitle?: string;
}

/**
 * Normalise text for MD5 hashing — mirrors the Python pipeline approach.
 * Lowercases, strips punctuation, collapses whitespace.
 */
export function normaliseTextForHash(text: string): string {
  return text
    .toLowerCase()
    .trim()
    .replace(/[^\w\s]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Check for exact content duplicates by comparing MD5 hashes of normalised content.
 * Computes the hash client-side for the input text, then uses the find_exact_duplicates
 * RPC to match against the stored content_text_hash generated column.
 * No content text is transferred -- only the 32-char hex hash is sent to the DB.
 */
async function findExactDuplicates(
  supabase: SupabaseClient<Database>,
  contentText: string,
  excludeId?: string,
): Promise<DuplicateMatch[]> {
  const normalised = normaliseTextForHash(contentText);
  if (!normalised) return [];

  const crypto = await import('crypto');
  const hash = crypto.createHash('md5').update(normalised).digest('hex');

  const { data: results, error } = await supabase.rpc('find_exact_duplicates', {
    p_content_hash: hash,
    p_exclude_id: excludeId ?? undefined,
  });

  if (error || !results) {
    logger.error({ err: error }, 'Exact dedup query failed');
    return [];
  }

  return (results as Array<{ id: string; title: string }>).map((r) => ({
    id: r.id,
    title: r.title ?? 'Untitled',
    similarity: 1.0,
    match_type: 'exact' as const,
  }));
}

/**
 * Check for near-duplicate content using embedding cosine similarity.
 * Uses the find_similar_content RPC which leverages pgvector HNSW indexes.
 */
async function findNearDuplicates(
  supabase: SupabaseClient<Database>,
  embedding: number[],
  threshold: number,
  excludeId?: string,
): Promise<DuplicateMatch[]> {
  const { data: results, error } = await supabase.rpc('find_similar_content', {
    query_embedding: JSON.stringify(embedding),
    similarity_threshold: threshold,
    limit_count: MAX_NEAR_DUPLICATE_RESULTS + (excludeId ? 1 : 0),
  });

  if (error || !results) {
    logger.error({ err: error }, 'Near-duplicate search failed');
    return [];
  }

  return results
    .filter((r: { id: string }) => r.id !== excludeId)
    .slice(0, MAX_NEAR_DUPLICATE_RESULTS)
    .map((r: { id: string; title: string; similarity: number }) => ({
      id: r.id,
      title: r.title ?? 'Untitled',
      similarity: r.similarity,
      match_type: 'near_duplicate' as const,
    }));
}

/**
 * Run full deduplication check on content text.
 *
 * Performs both exact hash matching and near-duplicate embedding search.
 * Returns matches sorted by similarity (exact matches first at 1.0).
 *
 * @param supabase - Authenticated Supabase client
 * @param contentText - The content text to check for duplicates
 * @param embedding - Pre-computed embedding vector (if available)
 * @param options - Configuration options
 * @returns DedupResult with any matches found
 */
export async function checkForDuplicates(
  supabase: SupabaseClient<Database>,
  contentText: string,
  embedding?: number[],
  options: {
    /** Cosine similarity threshold for near-duplicate detection (0.0-1.0) */
    nearDuplicateThreshold?: number;
    /** Exclude this item ID from results (useful when updating existing items) */
    excludeId?: string;
  } = {},
): Promise<DedupResult> {
  const threshold =
    options.nearDuplicateThreshold ?? DEFAULT_NEAR_DUPLICATE_THRESHOLD;
  const allMatches: DuplicateMatch[] = [];

  // 1. Exact match check
  try {
    const exactMatches = await findExactDuplicates(
      supabase,
      contentText,
      options.excludeId,
    );
    allMatches.push(...exactMatches);
  } catch (err) {
    logger.error({ err }, 'Exact dedup check failed');
  }

  // 2. Near-duplicate check (only if embedding available)
  if (embedding) {
    try {
      const nearMatches = await findNearDuplicates(
        supabase,
        embedding,
        threshold,
        options.excludeId,
      );
      // Exclude items already found as exact matches
      const exactIds = new Set(allMatches.map((m) => m.id));
      for (const match of nearMatches) {
        if (!exactIds.has(match.id)) {
          allMatches.push(match);
        }
      }
    } catch (err) {
      logger.error({ err }, 'Near-duplicate check failed');
    }
  }

  // Sort: exact matches first, then by similarity descending
  allMatches.sort((a, b) => {
    if (a.match_type === 'exact' && b.match_type !== 'exact') return -1;
    if (a.match_type !== 'exact' && b.match_type === 'exact') return 1;
    return b.similarity - a.similarity;
  });

  return {
    has_duplicates: allMatches.length > 0,
    matches: allMatches,
  };
}

/**
 * Shared dedup gate — returns the first exact-hash match (if any).
 *
 * Used by every TS entry point that inserts into `content_items` so
 * duplicate content is caught consistently. Soft-block contract:
 * callers should proceed with the insert and stamp
 * `dedup_status = 'suspected_duplicate'` when `isDuplicate` is true.
 * Admins can override via `skip_dedup=true` at the caller.
 *
 * Reference: docs/specs/cross-system-dedup-spec.md §3.1
 */
export async function checkExactDuplicate(
  supabase: SupabaseClient<Database>,
  contentText: string,
  options: { excludeId?: string } = {},
): Promise<ExactDuplicateCheck> {
  const normalised = normaliseTextForHash(contentText);
  if (!normalised || normalised.length < DEDUP_MIN_CONTENT_LENGTH) {
    return { isDuplicate: false };
  }
  const matches = await findExactDuplicates(
    supabase,
    contentText,
    options.excludeId,
  );
  const first = matches[0];
  if (!first) return { isDuplicate: false };
  return {
    isDuplicate: true,
    existingId: first.id,
    existingTitle: first.title,
  };
}

/**
 * Resolve dedup stamp fields for an insert/update payload.
 *
 * Soft-block contract (spec §6 D1): callers proceed with the write and
 * stamp `dedup_status='suspected_duplicate'` + record the existing item
 * id in `metadata.suspected_duplicate_of` when an exact-hash match is
 * found. Admin-only `skipDedup=true` silently bypasses the stamp.
 *
 * Usage:
 *   const exact = dedupResult.matches.find(m => m.match_type === 'exact');
 *   const { dedup_status, suspected_duplicate_of } =
 *     resolveDedupStamp(exact?.id, { skipDedup });
 *   // set insertData.dedup_status + merge suspected_duplicate_of into metadata
 *
 * Reference: docs/specs/cross-system-dedup-spec.md §6 D1, D2
 */
export function resolveDedupStamp(
  existingId: string | undefined,
  options: { skipDedup?: boolean } = {},
): {
  dedup_status: 'clean' | 'suspected_duplicate';
  suspected_duplicate_of?: string;
} {
  if (options.skipDedup || !existingId) {
    return { dedup_status: 'clean' };
  }
  return {
    dedup_status: 'suspected_duplicate',
    suspected_duplicate_of: existingId,
  };
}

/**
 * Format dedup matches into a human-readable warning string.
 */
export function formatDedupWarning(result: DedupResult): string | null {
  if (!result.has_duplicates) return null;

  const exactCount = result.matches.filter(
    (m) => m.match_type === 'exact',
  ).length;
  const nearCount = result.matches.filter(
    (m) => m.match_type === 'near_duplicate',
  ).length;

  const parts: string[] = [];
  if (exactCount > 0) {
    parts.push(
      `${exactCount} exact duplicate${exactCount > 1 ? 's' : ''} found`,
    );
  }
  if (nearCount > 0) {
    parts.push(`${nearCount} near-duplicate${nearCount > 1 ? 's' : ''} found`);
  }

  const titles = result.matches
    .slice(0, 3)
    .map((m) => `"${m.title}" (${(m.similarity * 100).toFixed(0)}%)`)
    .join(', ');

  return `Potential duplicates: ${parts.join(', ')}. Matches: ${titles}`;
}
